### PR TITLE
openlineage, docs: tips for OpenLineage method implementation

### PR DESCRIPTION
**Pull request by mobuchowski**
Add tips for provider contributors on how to effectively add OpenLineage support.

### CODE REVIEW

**Review by pankajastro — APPROVED** [review]
LGTM

File: docs/apache-airflow-providers-openlineage/guides/developer.rst [hunk 1]
```diff
@@ -20,7 +20,7 @@
 Implementing OpenLineage in Operators
 -------------------------------------
 
-OpenLineage defines few methods for implementation for Operators.
+OpenLineage defines few methods for implementation for Operators. Those are referred to as OpenLineage methods.
```
Comment (phanikumv) [inline docs/apache-airflow-providers-openlineage/guides/developer.rst]:
```suggestion
OpenLineage defines few methods for implementation in Operators. Those are referred to as OpenLineage methods.
```

File: docs/apache-airflow-providers-openlineage/guides/developer.rst [hunk 2]
```diff
@@ -50,4 +49,25 @@ Instead of returning complete OpenLineage event, the provider defines ``Operator
...
+It allows OpenLineage consumers to properly match information about datasets gathered from different sources.
+
+Naming convention is described in the `OpenLineage docs <https://openlineage.io/docs/spec/naming>`__.
+
+OpenLineage implementation should not waste time of users that do not use it.
+This means not doing heavy processing or network calls in the ``execute`` method that aren't used by it.
+Better option is to save relevant information in Operator attribute - and then use it
+in OpenLineage method.
+
+Good example is ``BigQueryExecuteQueryOperator``. It saves ``job_ids`` of queries that were executer.
```
Comment (pankajkoti) [inline docs/apache-airflow-providers-openlineage/guides/developer.rst]:
```suggestion
Good example is ``BigQueryExecuteQueryOperator``. It saves ``job_ids`` of queries that were executed.
```

File: docs/apache-airflow-providers-openlineage/guides/developer.rst [hunk 3]
```diff
@@ -50,4 +49,25 @@ Instead of returning complete OpenLineage event, the provider defines ``Operator
...
+This allows users to use your provider even if they do not have OpenLineage provider installed.
+
+Another important point is to make sure your provider returns OpenLineage-compliant dataset names.
+It allows OpenLineage consumers to properly match information about datasets gathered from different sources.
+
+Naming convention is described in the `OpenLineage docs <https://openlineage.io/docs/spec/naming>`__.
+
+OpenLineage implementation should not waste time of users that do not use it.
+This means not doing heavy processing or network calls in the ``execute`` method that aren't used by it.
+Better option is to save relevant information in Operator attribute - and then use it
```
Comment (pankajkoti) [inline docs/apache-airflow-providers-openlineage/guides/developer.rst]:
```suggestion
Better option is to save relevant information in Operator attributes - and then use it
```

**Review by pankajkoti — APPROVED** [review]

File: docs/apache-airflow-providers-openlineage/guides/developer.rst (hunk 1 above)
Comment (uranusjr) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst]: It should be _a few_.

https://dictionary.cambridge.org/grammar/british-grammar/little-a-little-few-a-few

> (A) little and (a) few are quantifiers meaning ‘some’. Little and few have negative meanings. We use them to mean ‘not as much as may be expected or wished for’.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst [hunk 4]
```diff
@@ -50,4 +49,25 @@ Instead of returning complete OpenLineage event, the provider defines ``Operator
       run_facets: dict[str, BaseFacet] = Factory(dict)
       job_facets: dict[str, BaseFacet] = Factory(dict)
 
-OpenLineage integration takes care to enrich it with things like general Airflow facets, proper event time and type.
+OpenLineage integration itself takes care to enrich it with things like general Airflow facets, proper event time and type, creating proper OpenLineage RunEvent.
+
+How to properly implement OpenLineage methods?
+==============================================
+
+There are several things worth noting when implementing OpenLineage in operators.
```
Comment (uranusjr) [inline docs/apache-airflow-providers-openlineage/guides/developer.rst]:
```suggestion
There are a couple of things worth noting when implementing OpenLineage in operators.
```

I was anticipating a third thing when reading this, but there seems to be only two?

**Review by sunank200 — CHANGES_REQUESTED** [review]
A section would be useful about how one should add tests (both unit and system tests)

File: docs/apache-airflow-providers-openlineage/guides/developer.rst [hunk 5]
```diff
@@ -50,4 +49,25 @@ Instead of returning complete OpenLineage event, the provider defines ``Operator
...
+
+First, do not import OpenLineage methods on top-level, but in OL method itself.
+This allows users to use your provider even if they do not have OpenLineage provider installed.
+
```
Comment (phanikumv) [inline docs/apache-airflow-providers-openlineage/guides/developer.rst:61]: can we add an example here (or) link to an already existing example? The example will be useful for folks who are starting out on a new OL implementation

File: docs/apache-airflow-providers-openlineage/guides/developer.rst (hunk 4 above)
Comment (mobuchowski) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst]: Hmm, I was referring to everything below that line. I guess that's unclear, tried to fix this.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst (hunk 5 above)
Comment (mobuchowski) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst:61]: Added example from `GcsToGcsOperator` - due to it's simplicity.

**Comment by mobuchowski** [comment]
@sunank200 added section on writing tests. I assume system test PR will be merged in similar shape it's today, if not I'll update the section.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst (hunk 3 above)
Comment (mobuchowski) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst]: Thanks, fixed.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst (hunk 2 above)
Comment (mobuchowski) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst]: Thanks, fixed.

File: docs/apache-airflow-providers-openlineage/guides/developer.rst (hunk 1 above)
Comment (mobuchowski) [reply docs/apache-airflow-providers-openlineage/guides/developer.rst]: Thanks, fixed both mistakes.

**Pull request merged**